### PR TITLE
Feat/int 717 automatically detect git remote path in vscode

### PR DESCRIPTION
**Pull request by henoktsegayetoptal**
---

<details open="true"><summary>Generated summary (powered by <a href="https://app.graphite.dev">Graphite</a>)</summary>

> #### What Changed
> Added
> - DefaultConfigurationClass
> - extensionDependencies
> #### Testing
> Test that the extension works as expected by activating it and checking that the repository path is set correctly. Also, check that the extension dependencies are installed.
> #### Reasoning
> The DefaultConfigurationClass was added to handle the configuration of the extension, and extensionDependencies were added to ensure that the required dependencies are installed for the extension to work properly.
</details>

### CODE REVIEW

File: src/configuration.ts [hunk 1]
```diff
@@ -41,4 +43,76 @@ export const getConfiguration = () => {
...
+								'repositoryPath',
+								repositoryPath,
+								vscode.ConfigurationTarget.Global,
+							);
+					}
+				} catch (error) {
+					if (error instanceof Error) {
+						vscode.window.showErrorMessage(error.message);
+					}
+					console.error('error', error);
```
Comment (grzpab) [inline src/configuration.ts:72]: You can just write `console.error(error)`

File: src/configuration.ts [hunk 2]
```diff
@@ -41,4 +43,76 @@ export const getConfiguration = () => {
...
+		this.#messageBus.subscribe(MessageKind.extensionActivated, async () => {
+			const repositoryPathFromConfig = getConfiguration().repositoryPath;
+			if (!repositoryPathFromConfig) {
+				try {
+					const repositoryPath = await Promise.race([
+						this.getGitRemote(),
+						this.timeout(4000),
+					]);
+
+					if (repositoryPath) {
```
Comment (grzpab) [inline src/configuration.ts:59]: I'd revert the logic here (so return if `!repositoryPath`)

File: src/configuration.ts [hunk 3]
```diff
@@ -41,4 +43,76 @@ export const getConfiguration = () => {
...
+						vscode.window.showErrorMessage(error.message);
+					}
+					console.error('error', error);
+				}
+			}
+		});
+	}
+	getGitRemote = () =>
+		// eslint-disable-next-line no-async-promise-executor
+		new Promise(async (resolve, reject) => {
```
Comment (grzpab) [inline src/configuration.ts:79]: actually I think it's a serious antipattern: https://stackoverflow.com/questions/23803743/what-is-the-explicit-promise-construction-antipattern-and-how-do-i-avoid-it/23803744#23803744

I think you can make it an async function and create the promise within it when needed.

File: src/configuration.ts [hunk 4]
```diff
@@ -41,4 +43,76 @@ export const getConfiguration = () => {
...
+			const gitExtension = vscode.extensions.getExtension(
```
Comment (grzpab) [inline src/configuration.ts:80]: can it throw?

**Review by grzpab — CHANGES_REQUESTED** [review]
See inline comments

File: src/configuration.ts (hunk 4 above)
Comment (henoktsegayetoptal) [reply src/configuration.ts:78]: it can reject , not throw

**Review by grzpab — CHANGES_REQUESTED** [review]
See inline comments

File: src/configuration.ts [hunk 5]
```diff
@@ -41,4 +44,74 @@ export const getConfiguration = () => {
...
+
+			if (!gitExtension) {
+				return reject(new Error('Git extension not found'));
+			}
+			if (!gitExtension.isActive) {
+				gitExtension
+					.activate()
+					.then(() =>
+						this.getPathFromGitExtension(gitExtension, resolve),
+					);
```
[the remark this answers is not in the document]
Comment (henoktsegayetoptal) [reply src/configuration.ts:90]: unfortunately `activate()` doesn't have a catch

File: src/types/git.d.ts [hunk 6]
```diff
@@ -0,0 +1,209 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
```
[the remark this answers is not in the document]
Comment (grzpab) [reply src/types/git.d.ts:4]: did @DmytroHryshyn add the same file already?

**Review by grzpab — APPROVED** [review]

**Pull request merged**